### DISCUSSION TITLE
Migrate `getFormActions` from v2 to v3

### DISCUSSION THREAD

**Discussion by Roboroads** [filamentphp/filament#12675] [Help] [answered]
### Package

Panel builder

### Package Version

v3.2.72

### How can we help you?

We are migrating v2 to v3 and we came across the fact that it seems that `getFormActions` (on `App\Filament\Components\Page\CreateRecord`) gets executed _before_ an action that is defined within `getFormActions`. 

In v2 we did the following (It's an extremely simplified example):
```php
protected function getFormActions(): array
{
    if ($this->data['prepared']) {
        return parent::getFormActions();
    }

    return [
        Action::make('prepareForSave')
            ->action(function () { 
                // Doing some prep here...
                $this->data['prepared'] = true;
            ),
    ];
}
```
So, when clicking the prepareForSave, after some processing, the default buttons would show.

How do we migrate this to v3?

**Comment by leandrocfe** [comment]
Why not use `->visible($this->data['prepared'])` ?

**Reply by Roboroads** [reply]
As @EemeliSurakka mentioned, that doesn't fly anymore.

```php
protected function getFormActions(): array
{
    dump('getFormActions');
    if ($this->data['prepared']) {
        return parent::getFormActions();
    }

    return [
        Action::make('prepareForSave')
            ->action(function () { 
                // Doing some prep here...
                dump('prepareForSave');
                $this->data['prepared'] = true;
            ),
    ];
}
```

Pressing the action now would give you a dump that says:

- getFormActions
- prepareForSave

While in v2 this is the other way around.

Your suggestion would be something like this:
```php
protected function getFormActions(): array
{
    return [
        $this->getCreateFormAction()
            ->visible($this->data['prepared']),
        $this->getCreateAnotherFormAction()
            ->visible($this->data['prepared']),
        $this->getCancelFormAction()
            ->visible($this->data['prepared']),
        Action::make('prepareForSave')
            ->visible(!$this->data['prepared'])
            ->action(function () { 
                // Doing some prep here...
                dump('prepareForSave');
                $this->data['prepared'] = true;
            ),
    ];
}
```
The same problem occurs here, where when getFormActions gets ran, the `$this->data['prepared']` is not yet set to `true`.

**Reply by leandrocfe** [reply, marked as answer]
```php
private $prepared = false;

protected function getFormActions(): array
{
    return [
        ...parent::getFormActions(),
        Actions\Action::make('prepare')
            ->action(function () {
                $this->prepared = true;
            })
            ->hidden(fn () => $this->prepared),
    ];
}

protected function getSaveFormAction(): Action
{
    return parent::getSaveFormAction()
        ->visible(fn () => $this->prepared);
}

protected function getCancelFormAction(): Action
{
    return parent::getCancelFormAction()
        ->visible(fn () => $this->prepared);
}
```

**Reply by Roboroads** [reply]
Alright, the closure way works - we've rewritten it like that and works like a charm! Thanks! <3

**Comment by EemeliSurakka** [comment]
> Why not use `->visible($this->data['prepared'])` ?

This does not work, since when prepareForSave action is clicked the getFormActions method is executed again before the prepareForSave action callback. So the value of data['prepared'] is still false when actions are rendered again.
